proposals/NNNN-multi-pattern-and-conditionally-compiled-catches.md

```diff
@@ -0,0 +1,156 @@
+
+## Motivation
+
+Currently, Swift only allows up to one pattern and where clause for each catch clause in a do-catch statement, so the following code snippet is not allowed:
+
+```swift
+do {
+  try performTask()
+} catch TaskError.someRecoverableError {    // OK
+  recover()
+} catch TaskError.someFailure(let msg),
+        TaskError.anotherFailure(let msg) { // Not currently valid
+  showMessage(msg)
+}
+```
+
+Because the above snippet is not valid today, developers frequently end up duplicating code between catch clauses, or writing something like the following instead:
+
+```swift
+do {
+  try performTask()
+} catch {
+  switch error {
+  case TaskError.someRecoverableError:
+    recover()
+  case TaskError.someFailure(let msg),
+       TaskError.anotherFailure(let msg):
+    showMessage(msg)
+  }
+}
+```
+
+Nesting the switch inside of the catch clause is awkward and defeats the purpose of supporting pattern matching in catch clauses. Splitting the code up into multiple catch clauses requires duplicating the body, which is also undesirable. Supporting a multi-pattern catch clause would allow for code which is both clearer and more concise.
+
+The following use of `#if` is also not allowed today and frequently leads to nesting of switch statements inside catch clauses:
+
+```swift
+do {
+  try performTask()
+}
+#if os(macOS)
+catch TaskError.macOSOnlyError {
+  macOSRecovery()
+}
+#endif
+catch { 
+  crossPlatformRecovery()
+}
+``` 
+
+## Proposed solution
+
+Catch clauses should allow the user to specify a comma-separated list of patterns. If an error thrown at runtime in a do block matches any of the patterns in a corresponding catch clause, that catch clause's body should be executed. Similar to switch cases, a user should be able to bind a variable in all patterns of a catch clause and then use it in the body.
+
+Additionally, conditional compilation directives should be permitted to appear surrounding catch clauses.
+
+When used together, these additions allow writing code like the following:
+
+```swift
+do {
+  try performTask()
+} 
+#if os(macOS)
+catch TaskError.macOSOnlyError {            // A conditionally compiled catch clause
+  macOSRecovery()
+}
+#endif
+catch TaskError.someOtherFailure(let msg),  // Multiple patterns in a single catch clause
+      TaskError.anotherFailure(let msg) {
+  showMessage(msg)
+}
+```
+
+## Detailed design
+
+### Grammar Changes
+
+The revised catch clause grammar is as follows:
+
+```
+catch-clauses -> catch-clause catch-clauses?
+
+catch-clause -> 'catch' catch-item-list? code-block |
+                conditional-catch-clause
+
+catch-item-list -> catch-item |
+                   catch-item ',' case-item-list
+
+catch-item -> pattern where-clause? |
+              where-clause
+
+conditional-catch-clause -> catch-if-directive-clause catch-elseif-directive-clauses? switch-else-directive-clause? endif-directive
```

Comment:
```suggestion
conditional-catch-clause -> catch-if-directive-clause catch-elseif-directive-clauses? catch-else-directive-clause? endif-directive
```

Reply:
Thanks, fixed that and another typo in the grammar.